### PR TITLE
[ARTS-286] Role service

### PR DESCRIPTION
**Pull request by KateSant**
## Description

NB this is the CLEAN branch (avoiding the rewritten git history issues).

New role-service.   Plus refactor of trial-config-service to make it call role-service and associated docker compose.

### Changes
task [ARTS-286] / story [ARTS-98]

[ARTS-286]: https://ndph-arts.atlassian.net/browse/ARTS-286
[ARTS-98]: https://ndph-arts.atlassian.net/browse/ARTS-98

### CODE REVIEW

File: docker-compose.yml
Comment (mdp0) [inline docker-compose.yml:31]: Shouldn't this just be the service name, rather than container? I.e. role-service:82

File: role-service/pom.xml
Comment (tamirkamara) [inline role-service/pom.xml:91]: This exists in the parent...

**Review by tamirkamara — CHANGES_REQUESTED** [review]

File: docker-compose.yml
Comment (KateSant) [reply docker-compose.yml]: Thanks Mark, good point. I have changed this.

File: role-service/pom.xml
Comment (KateSant) [reply role-service/pom.xml]: Thanks Tamir, good point.  I have cleaned up this pom.

File: build-tools/src/main/resources/checkstyle-suppressions.xml
Comment (KateSant) [reply build-tools/src/main/resources/checkstyle-suppressions.xml]: Thanks Tamir good point.   I have removed these.  Same for the trial-config-service suppressions I think?

File: role-service/pom.xml
Comment (KateSant) [reply role-service/pom.xml]: Thanks Tamir, good point.  I have cleaned up this pom.

File: role-service/src/main/java/uk/ac/ox/ndph/mts/role_service/controller/RoleController.java
Comment (KateSant) [reply role-service/src/main/java/uk/ac/ox/ndph/mts/role_service/controller/RoleController.java:22]: I'm not sure - there is a debate on singular vs plural convention for REST endpoints.  We generally use plural, see: https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design   But I do not have a strong view - happy to change to singular if you want?  We should be consistent across our services though.

Comment (tamirkamara) [reply role-service/src/main/java/uk/ac/ox/ndph/mts/role_service/controller/RoleController.java:22]: You're right. So we'll need to adjust the other services (something for another day)

File: build-tools/src/main/resources/checkstyle-suppressions.xml
Comment (tamirkamara) [reply build-tools/src/main/resources/checkstyle-suppressions.xml]: Right. you can delete those...

File: role-service/src/test/java/uk/ac/ox/ndph/mts/role_service/controller/RoleControllerTest.java
Comment (tamirkamara) [inline role-service/src/test/java/uk/ac/ox/ndph/mts/role_service/controller/RoleControllerTest.java:63]: Do we need to validate that the role name is unique?

File: role-service/src/main/java/uk/ac/ox/ndph/mts/role_service/model/Role.java
Comment (tamirkamara) [inline role-service/src/main/java/uk/ac/ox/ndph/mts/role_service/model/Role.java:29]: This means that we can have roles defined with a specific ID as described in the trial config example.  I don't have a specific ask here...

Comment (KateSant) [reply role-service/src/main/java/uk/ac/ox/ndph/mts/role_service/model/Role.java:29]: Yes.  I wonder whether it makes more sense to reference them by (unique) role name in the json, instead of a generated int?

File: role-service/src/test/java/uk/ac/ox/ndph/mts/role_service/controller/RoleControllerTest.java
Comment (KateSant) [reply role-service/src/test/java/uk/ac/ox/ndph/mts/role_service/controller/RoleControllerTest.java:63]: I am not sure - maybe the ID is the unique property to rely on?   OR if the name is the true unique ID, then perhaps we don't need the numeric ID?   I am in discussion with Adrian about the various properties of roles and permissions today, so I will raise this with him and I'll deal with this in branch 101 if it's to be unique.  OK to leave as-is for this PR?

**Review by tamirkamara — DISMISSED** [review]

File: role-service/src/test/java/uk/ac/ox/ndph/mts/role_service/controller/RoleControllerTest.java
Comment (KateSant) [reply role-service/src/test/java/uk/ac/ox/ndph/mts/role_service/controller/RoleControllerTest.java:67]: Change of plan - I have confirmed that the name is unique and we can use it as the ID, so I've added the validation and error handling to pick up duplicates.

File: trial-config-service/pom.xml
Comment (shajidissa) [inline trial-config-service/pom.xml:114]: Kate can you remove this dependency please

File: role-service/pom.xml
Comment (shajidissa) [inline role-service/pom.xml:96]: Kate can you remove the lombok dependency please

Comment (tamirkamara) [reply role-service/pom.xml:96]: @shajidissa, @KateSant has another task about permissions on the role-service which is starting right after this. I think we should approve this one and remove lombok as part of the other task. wdyt?

**Review by tamirkamara — DISMISSED** [review]

File: role-service/pom.xml
Comment (KateSant) [reply role-service/pom.xml:96]: Why are we dropping lombok?  (It has come into fashion and gone out again in the time this feature has been in dev!)  Yes please I agree with @tamirkamara I think it would be better as a separate task across all services, I can then unblock Sameera to get on with QA on roles.

**Review by mdp0 — APPROVED** [review]

**Review by tamirkamara — APPROVED** [review]

File: role-service/pom.xml
Comment (KateSant) [reply role-service/pom.xml:96]: Discussed in another thread with @shajidissa and agreed.

**Pull request merged**
